Hyprlinkr/DefaultRouteDispatcher.cs

```diff
@@ -108,5 +95,73 @@ public string RouteName
+                        
+            var controllerRouteAttribute = controllerType.GetCustomAttribute<RouteAttribute>(false);
+            if (controllerRouteAttribute != null)
+            {
+                return ExtractRoupleFromControllerRouteAttribute(callExpression, routeValues, controllerRouteAttribute);
+            }
+
+            var controllerName = controllerType.Name.ToLowerInvariant().Replace("controller", "");
+
+            var newRouteValues = new Dictionary<string, object>(routeValues);
+            newRouteValues["controller"] = controllerName;
+
+            return new Rouple(this.routeName, newRouteValues);
+        }
+
+        private Rouple ExtractRoupleFromControllerRouteAttribute(
+            MethodCallExpression callExpression, 
+            IDictionary<string, object> routeValues, 
+            RouteAttribute controllerRouteAttribute)
+        {
+            var controllerRouteName = this.routeName;
+            if (controllerRouteAttribute.Name != null)
+            {
+                controllerRouteName = controllerRouteAttribute.Name;
+            }
+
+            var newRouteValues = AddControllerNameAndMethodToRouteValues(callExpression, routeValues);
+            newRouteValues = RemoveControllerAndActionRouteValuesIfNeeded(controllerRouteAttribute, newRouteValues);
+            return new Rouple(controllerRouteName, newRouteValues);
+        }
+
+        private static Dictionary<string, object> RemoveControllerAndActionRouteValuesIfNeeded(
+            RouteAttribute controllerRouteAttribute, 
+            Dictionary<string, object> routeValues)
+        {
+            var controllerRouteTemplate = controllerRouteAttribute.Template;
+            if (controllerRouteTemplate != null)
+            {
+                if (!controllerRouteTemplate.Contains("{controller}"))
+                {
+                    routeValues.Remove("controller");
+                }
+                if (!controllerRouteTemplate.Contains("{action}"))
+                {
+                    routeValues.Remove("action");
+                }
+            }
+            return routeValues;
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Globalization", "CA1308:NormalizeStringsToUppercase", Justification = "This method should produce URIs with lower-case letters, so ultimately, it would have to invoke some sort of ToLower method.")]
+        private static Dictionary<string, object> AddControllerNameAndMethodToRouteValues(
+            MethodCallExpression callExpression, 
+            IDictionary<string, object> routeValues)
+        {
+            var newRouteValues = new Dictionary<string, object>(routeValues);
+            var controllerName = callExpression.Object.Type.Name.ToLowerInvariant().Replace("controller", "");
+            newRouteValues["controller"] = controllerName;
+            newRouteValues["action"] = callExpression.Method.Name.ToLowerInvariant();
```

Comment:
Here's a suggestion, which will [make things go smoother](http://blog.ploeh.dk/2015/01/15/10-tips-for-better-pull-requests):

1. Decline this pull request. You can do this yourself.
2. Send a new pull request without these extra features, but only handling of Controller-level attributes.
3. Send _another_ pull request later with the other changes.

Making pull requests smaller tend to make the entire process faster.